### PR TITLE
Various tag improvements

### PR DESCRIPTION
**Pull request by jpzwarte**
`<sl-tag>`:
- Simplify the HTML structure
- Slot the text instead of using a property
- Use a `ResizeObserver` for determining whether the text overflows
- Rework the CSS so we don't need JS for the hover styling
- Use a lazy tooltip
- Remove `readonly` property for now due to uncertainty wrt focusability (https://adrianroselli.com/2022/11/brief-note-on-aria-readonly-support-html.html)

`<sl-tag-list>`:
- Remove `MutationObserver` and watch the tags using the `ResizeObserver` instead
- Simplify styling by setting a class on the `stack`

`Tooltip.lazy`: returns a cleanup function that can be used to clean up, but also to prevent multiple lazy tooltips from being initialized.

### CODE REVIEW

File: packages/components/tag/src/tag.stories.ts [hunk 1]
```diff
@@ -1,292 +1,121 @@
...
-        ></sl-tag>
-        ${amount.map(
-          (_, index) => html`
-            <sl-tag .label=${label + ' ' + (index + 1)} ?removable=${removable} ?readonly=${readonly}></sl-tag>
-          `
-        )}
-      </sl-tag-list>
-    `;
+    label: 'This is a very long label which overflows',
+    maxWidth: '200px'
```
Comment (anna-lach) [inline packages/components/tag/src/tag.stories.ts:62]: When I change the maxWidth in the story, even when there is ellipsis visible, the tooltip is no longer showing on hover/focus

image

File: packages/components/tag/src/tag.scss [hunk 2]
```diff
@@ -1,263 +1,134 @@
...
-
-  .remove-button {
-    background: var(--sl-color-tag-subtle-hover-background);
-
-    --_button-border: var(--sl-color-tag-subtle-hover-border);
-  }
+slot:hover,
+slot:hover + button,
+button:hover {
+  --_background: var(--sl-color-tag-subtle-hover-background);
```
Comment (anna-lach) [inline packages/components/tag/src/tag.scss:28]: When the bold tag is focused and I hover it (the tag, not the button in it), the background of the button changes, but it should not when I hover over the tag, not the button. Right now it looks like this:

image

File: packages/components/tag/src/tag-list.scss [hunk 3]
```diff
@@ -1,89 +1,76 @@
...
-    }
+  &:has(sl-tag[emphasis='bold']:focus) {
+    --_border-color: var(--sl-color-tag-subtle-active-border);
   }
 }
 
+.double::before,
+.triple::before,
+.triple::after {
+  border-color: var(--_border-color);
```
Comment (anna-lach) [inline packages/components/tag/src/tag-list.scss:56]: It looks like it's not alligned with the tag list?
image
It sticks out from the left side?

File: packages/components/tag/src/tag-list.scss [hunk 4]
```diff
@@ -1,89 +1,76 @@
...
+  border-inline-start: 2px solid var(--_border-color);
```
Comment (anna-lach) [inline packages/components/tag/src/tag-list.scss:57]: Why we are not using tokens here? :)

File: packages/components/tag/src/tag-list.scss [hunk 5]
```diff
@@ -1,89 +1,76 @@
...
+  border-radius: 4px;
```
Comment (anna-lach) [inline packages/components/tag/src/tag-list.scss:58]: Can you please use tokens here? :)

File: packages/components/tag/src/tag-list.scss [hunk 6]
```diff
@@ -1,89 +1,76 @@
...
+  content: '';
+  display: flex;
+  inline-size: 4px;
```
Comment (anna-lach) [inline packages/components/tag/src/tag-list.scss:61]: Can you please use tokens here? :)

File: packages/components/tag/src/tag-list.scss (hunk 5 above)
Comment (anna-lach) [reply packages/components/tag/src/tag-list.scss:58]: We could have problems with shape in other themes when we will leave the exact value here, not a token?

File: packages/components/tag/src/tag-list.ts [hunk 7]
```diff
@@ -47,175 +41,161 @@ export class TagList extends ScopedElementsMixin(LitElement) {
...
-      this.tags?.forEach(tag => (tag.emphasis = this.emphasis));
+    if (changes.has('stacked') && !this.stacked) {
+      this.tags.forEach(tag => (tag.style.display = ''));
+    }
+
+    if (changes.has('stacked') || changes.has('stackSize') || changes.has('tags')) {
+      if (this.stacked) {
+        const total = this.tags.length;
+
+        this.setAttribute('aria-label', `${msg(str`Showing ${total - this.stackSize} out of ${total} tags`)}`);
```
Comment (anna-lach) [inline packages/components/tag/src/tag-list.ts:118]: So we are leaving `tags` for now, yes?

File: packages/components/tag/src/tag.scss [hunk 8]
```diff
@@ -1,263 +1,134 @@
...
-  border-inline-end: none;
-  border-inline-start: 1px solid var(--_button-border);
-  border-start-end-radius: var(--_border-radius);
-  cursor: pointer;
-  display: flex;
-  inline-size: var(--sl-size-tag-md);
+  aspect-ratio: 1;
+  background: var(--_background);
+  border: 0;
+  border-inline-start: 1px solid var(--_border-color);
```
Comment (anna-lach) [inline packages/components/tag/src/tag.scss:121]: Can you please use a token for the border width?

File: packages/components/tag/src/tag.stories.ts (hunk 1 above)
Comment (jpzwarte) [reply packages/components/tag/src/tag.stories.ts:62]: I cannot reproduce that at this time.

Comment (anna-lach) [reply packages/components/tag/src/tag.stories.ts:62]: I think when I changed the `maxWidth` few times there in options it happened.

File: packages/components/tag/src/tag-list.scss (hunk 3 above)
Comment (jpzwarte) [reply packages/components/tag/src/tag-list.scss:56]: Fixed

File: packages/components/tag/src/tag.ts [hunk 9]
```diff
@@ -58,128 +65,105 @@ export class Tag extends ScopedElementsMixin(LitElement) {
...
+  }
 
-    this.#checkOverflow();
+  override render(): TemplateResult {
+    return html`
+      <div class="wrapper" .tabIndex=${this.disabled ? -1 : 0}>
+        <slot @slotchange=${this.#onSlotChange}></slot>
+        ${this.removable
+          ? html`
+              <button @click=${this.#onRemove} aria-label=${msg(`Remove ${this.label}`)} ?disabled=${this.disabled}>
```
Comment (anna-lach) [inline packages/components/tag/src/tag.ts:119]: It looks like the `this.label` is not being rendered inside the `aria-label`?

image

File: packages/components/tag/src/tag.ts [hunk 10]
```diff
@@ -58,128 +65,105 @@ export class Tag extends ScopedElementsMixin(LitElement) {
...
+                <sl-icon name="xmark" size=${ifDefined(this.size)}></sl-icon>
```
Comment (anna-lach) [inline packages/components/tag/src/tag.ts:120]: I've just noticed that for `xmark` icon in Figma there is a `xs` size of icon used for `md` and `lg` size of tag,
 so 12x12 px:

image

**Review by anna-lach — APPROVED** [review]

**Review by Diaan — APPROVED** [review]

**Pull request merged**